astropy/coordinates/sky_coordinate.py

```diff
@@ -484,6 +487,94 @@ def transform_to(self, frame, merge_attributes=True):
+
+        Parameters
+        ----------
+        new_obstime : `~astropy.time.Time`, optional
+            The time at which to evolve the position to.
+        dt : `~astropy.units.Quantity`, `~astropy.time.TimeDelta`, optional
+            An amount of time to evolve the position of the source.
+        """
+
+        if (new_obstime is None and dt is None or
+                new_obstime is not None and dt is not None):
+            raise ValueError("You must specify one of `new_obstime` or `dt`, "
+                             "but not both.")
+
+        # Validate that we have velocity info
+        if 's' not in self.frame.data.differentials:
+            raise ValueError('SkyCoord requires velocity data to evolve the '
+                             'position.')
+
+        if 'obstime' in self.frame.frame_attributes:
+            raise NotImplementedError("Updating the coordinates in a frame "
+                                      "with explicit time dependence is "
+                                      "currently not supported. If you would "
+                                      "like this functionality, please open an "
+                                      "issue on github:\n"
+                                      "https://github.com/astropy/astropy")
+
+        if new_obstime is not None and self.obstime is None:
+            # If no obstime is already on this object, raise an error if a new
+            # obstime is passed: we need to know the time / epoch at which the
+            # the position / velocity were measured initially
+            raise ValueError('This object has no associated `obstime`. '
+                             'apply_space_motion() must receive a time '
+                             'difference, `dt`, and not a new obstime.')
+
+        t1 = self.obstime
+        if t1 is None:
+            # MAGIC NUMBER: if the current SkyCoord object has no obstime,
+            # assume J2000 to do the dt offset.
+            t1 = Time('J2000')
+
+        if dt is None:
+            dt = new_obstime - self.obstime
+        t2 = t1 + dt
+
+        icrs = self.icrs
+        icrs.set_representation_cls(s=SphericalDifferential)
+
+        try:
+            plx = icrs.distance.to(u.arcsecond, u.parallax()).value
+        except u.UnitConversionError: # No distance: set to 0 by starpm convention
+            plx = 0.
+
+        try:
+            rv = icrs.radial_velocity.to(u.km/u.s).value
+        except u.UnitConversionError: # No RV
+            rv = 0.
+
+        # proper motion in RA should not include the cos(dec) term, see the
+        # erfa function eraStarpv, comment (4).
+        starpm = erfa.starpm(icrs.ra.radian, icrs.dec.radian,
+                             icrs.pm_ra.to(u.radian/u.yr).value,
+                             icrs.pm_dec.to(u.radian/u.yr).value,
+                             plx, rv,
+                             2400000.5, t1.tdb.mjd,
```

Comment:
We have a utility for this:
```
# somewhere on top
from .builtin_frames.utils import get_jd12
# then here
*get_jd12(old_obstime, 'tdb'),
*get_jd12(new_obstime, 'tdb'))
```

Reply:
I'm not familiar with the `jd1`/`jd2` formality in time, so just want to make sure: is the output of `get_jd12()` compatible with the input of `starpm()`?

```
**  1) The starting and ending TDB dates ep1a+ep1b and ep2a+ep2b are
**     Julian Dates, apportioned in any convenient way between the two
**     parts (A and B).  For example, JD(TDB)=2450123.7 could be
**     expressed in any of these ways, among others:
**
**             epna          epnb
**
**         2450123.7           0.0       (JD method)
**         2451545.0       -1421.3       (J2000 method)
**         2400000.5       50123.2       (MJD method)
**         2450123.5           0.2       (date & time method)
```

Reply:
Yes, it is meant to feed jd1,jd2 to erfa.